File: shared/contracts/src/imports.ts
export type PlaylistPlatform = "netease" | "qq";

export type PlaylistPreviewStatus = "ready" | "failed";

export interface ImportPreviewRequest {
  raw_share_texts: string[];
}

export interface PlaylistPreviewError {
  code: string;
  message: string;
}

export interface PlaylistPreviewItem {
  platform?: PlaylistPlatform;
  canonical_url?: string;
  source_playlist_id?: string;
  title?: string;
  owner_source_id?: string;
  owner_nickname?: string;
  owner_avatar_url?: string;
  cover_url?: string;
  source_tags?: string[];
  track_count?: number;
  preview_status: PlaylistPreviewStatus;
  error?: PlaylistPreviewError;
}

export interface ImportPreviewResponse {
  items: PlaylistPreviewItem[];
}

export interface ConfirmedSourcePlaylist {
  platform: PlaylistPlatform;
  canonical_url: string;
  source_playlist_id: string;
  title: string;
  owner_source_id: string;
  owner_nickname: string;
  owner_avatar_url?: string;
  cover_url?: string;
  source_tags?: string[];
  track_count?: number;
  import_track_limit?: number;
}

export interface FullImportRequest {
  source_playlists: ConfirmedSourcePlaylist[];
}

export interface SourceTrackItem {
  id: string;
  platform: PlaylistPlatform;
  source_track_id: string;
  title: string;
  artists: string[];
  album?: string;
  duration_ms?: number;
  cover_url?: string;
  source_playlist_id: string;
  owner_source_id: string;
  owner_nickname: string;
  owner_avatar_url?: string;
}

export type SourcePlaylistImportStatus =
  | "pending"
  | "reading"
  | "ready"
  | "failed";

export type ImportSessionStatus =
  | "pending"
  | "reading"
  | "ready"
  | "partial_failed"
  | "failed"
  | "previewed"
  | "importing"
  | "normalizing"
  | "matching"
  | "syncing_temp"
  | "ready_to_play";

export type ImportFailedStage =
  | "importing"
  | "normalizing"
  | "matching"
  | "syncing_temp";

export type AnalyticsStatus =
  | "pending"
  | "running"
  | "partial"
  | "completed"
  | "failed";

export interface SourcePlaylistImportResult extends ConfirmedSourcePlaylist {
  id: string;
  status: SourcePlaylistImportStatus;
  read_count: number;
  error?: PlaylistPreviewError;
}

export interface ImportSessionResponse {
  id: string;
  status: ImportSessionStatus;
  raw_track_count: number;
  source_playlists: SourcePlaylistImportResult[];
  tracks: SourceTrackItem[];
  created_at: string;
  updated_at: string;
  failed_stage?: ImportFailedStage;
  error?: PlaylistPreviewError;
  progress?: ImportProgress;
  temp_playlist_id?: string;
  ready_to_play_at?: string;
  analytics_job_id?: string;
  analytics_status?: AnalyticsStatus;
  analytics_results: AnalyticsMetric[];
  matched_tracks: MatchedTrackItem[];
  playback?: ImportPlaybackPayload;
}

export interface ImportStageProgress {
  current: number;
  total: number;
}

export interface ImportProgress {
  read: ImportStageProgress;
  match: ImportStageProgress;
  sync: ImportStageProgress;
}

export interface AnalyticsMetric {
  metric_key: string;
  payload: Record<string, unknown>;
  status: string;
  computed_at: string;
}

export interface ImportPlaybackPayload {
  temp_playlist_id: string;
  tracks: MatchedTrackItem[];
}

export interface ImportHistorySourceSummary {
  platform: PlaylistPlatform;
  source_playlist_id: string;
  title: string;
  owner_nickname: string;
  import_track_limit?: number;
  read_count: number;
}

export interface ImportHistoryItem {
  session_id: string;
  ready_to_play_at: string;
  temp_playlist_id: string;
  playable_track_count: number;
  source_playlists: ImportHistorySourceSummary[];
  owner_nicknames: string[];
  created_at: string;
  updated_at: string;
}

export interface ImportSessionDeleteResponse {
  session_id: string;
  deleted: boolean;
}

export interface Contributor {
  platform: PlaylistPlatform;
  source_playlist_id: string;
  owner_source_id: string;
  owner_nickname: string;
  owner_avatar_url?: string;
}

export interface UnifiedTrackItem {
  id: string;
  normalized_title: string;
  display_title: string;
  artists: string[];
  normalized_artists: string[];
  album?: string;
  normalized_album?: string;
  duration_ms?: number;
  cover_url?: string;
  source_track_ids: string[];
  contributors: Contributor[];
  explain_dedup_reason: string;
}

export interface DedupeTracksResponse {
  import_session_id: string;
  raw_track_count: number;
  unique_track_count: number;
  tracks: UnifiedTrackItem[];
}

export interface NeteaseTrackCandidate {
  netease_song_id: string;
  title: string;
  artists: string[];
  album?: string;
  duration_ms?: number;
  score: number;
  reason: string;
}

export interface GenreAssignment {
  genre: string;
  source: string;
  confidence: number;
}

export type MatchStatus =
  | "auto_accepted"
  | "needs_confirm"
  | "no_match"
  | "manual_confirmed";

export interface MatchedTrackItem {
  id: string;
  display_title: string;
  artists: string[];
  album?: string;
  duration_ms?: number;
  cover_url?: string;
  source_track_ids: string[];
  contributors: Contributor[];
  genre_assignments?: GenreAssignment[];
  match_status: MatchStatus;
  netease_song_id?: string;
  match_confidence: number;
  match_reason: string;
  candidates: NeteaseTrackCandidate[];
}

export interface MatchTracksResponse {
  import_session_id: string;
  total_track_count: number;
  auto_matched_count: number;
  needs_confirm_count: number;
  no_match_count: number;
  tracks: MatchedTrackItem[];
}

export type MatchJobStatus =
  | "pending"
  | "running"
  | "ready"
  | "failed"
  | "rate_limited";

export interface MatchJobResponse {
  id: string;
  import_session_id: string;
  status: MatchJobStatus;
  processed_track_count: number;
  total_track_count: number;
  auto_matched_count: number;
  needs_confirm_count: number;
  no_match_count: number;
  current_title?: string;
  error?: string;
  result?: MatchTracksResponse;
}

export interface ManualMatchConfirmRequest {
  source_track_ids: string[];
  netease_song_id: string;
  title: string;
  artists: string[];
  album?: string;
  duration_ms?: number;
}

export type TempPlaylistSyncStatus = "ready" | "partial_failed";

export interface TempPlaylistBatchResult {
  operation: "add" | "remove";
  start_index: number;
  track_count: number;
  attempt: number;
  status: "ok" | "failed";
  error?: string;
}

export interface TempPlaylistSyncResponse {
  import_session_id: string;
  temp_playlist_id: string;
  status: TempPlaylistSyncStatus;
  synced_count: number;
  skipped_count: number;
  failed_count: number;
  ready_at?: string;
  batches: TempPlaylistBatchResult[];
  error?: string;
}

export interface KnownTempPlaylistSyncRequest {
  import_session_id: string;
  netease_song_ids: string[];
}

File: services/api/app/schemas/imports.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

from app.schemas.matching import MatchedTrackItem


class ImportPreviewRequest(BaseModel):
    raw_share_texts: list[str] = Field(min_length=1, max_length=20)


class PlaylistPreviewError(BaseModel):
    code: str
    message: str


class PlaylistPreviewItem(BaseModel):
    platform: Literal["netease", "qq"] | None = None
    canonical_url: str | None = None
    source_playlist_id: str | None = None
    title: str | None = None
    owner_source_id: str | None = None
    owner_nickname: str | None = None
    owner_avatar_url: str | None = None
    cover_url: str | None = None
    source_tags: list[str] = Field(default_factory=list)
    track_count: int | None = None
    preview_status: Literal["ready", "failed"]
    error: PlaylistPreviewError | None = None


class ImportPreviewResponse(BaseModel):
    items: list[PlaylistPreviewItem]


class ConfirmedSourcePlaylist(BaseModel):
    platform: Literal["netease", "qq"]
    canonical_url: str
    source_playlist_id: str
    title: str
    owner_source_id: str
    owner_nickname: str
    owner_avatar_url: str | None = None
    cover_url: str | None = None
    source_tags: list[str] = Field(default_factory=list)
    track_count: int | None = None
    import_track_limit: int | None = Field(
        default=None,
        ge=1,
        description=(
            "Optional per-source playlist import cap. When set, the backend "
            "keeps a deterministic random sample of at most this many tracks "
            "from this source playlist; when omitted, all fetched tracks are imported."
        ),
    )


class FullImportRequest(BaseModel):
    source_playlists: list[ConfirmedSourcePlaylist] = Field(min_length=1, max_length=20)


class SourceTrackItem(BaseModel):
    id: str
    platform: Literal["netease", "qq"]
    source_track_id: str
    title: str
    artists: list[str]
    album: str | None = None
    duration_ms: int | None = None
    cover_url: str | None = None
    source_playlist_id: str
    owner_source_id: str
    owner_nickname: str
    owner_avatar_url: str | None = None


class SourcePlaylistImportResult(ConfirmedSourcePlaylist):
    id: str
    status: Literal["pending", "reading", "ready", "failed"]
    read_count: int = 0
    error: PlaylistPreviewError | None = None


class ImportStageProgress(BaseModel):
    current: int = 0
    total: int = 0


class ImportProgress(BaseModel):
    read: ImportStageProgress = Field(default_factory=ImportStageProgress)
    match: ImportStageProgress = Field(default_factory=ImportStageProgress)
    sync: ImportStageProgress = Field(default_factory=ImportStageProgress)


class AnalyticsMetric(BaseModel):
    """Persisted Analytics V1/V2 metric.

    Formulas use SourcePlaylist owner_source_id, track contributors, and
    unique UnifiedTrack/MatchedTrack rows. Playback history is never used.
    Jaccard metrics use intersection / union and are rounded to 6 decimals.
    Exclusive ratios use exclusive_count / owner_total and are rounded to
    6 decimals. The owner pair score is the mean of track_jaccard and
    artist_jaccard, with component numbers included in the payload.

    Analytics V2 genre formulas never infer style with an LLM. GenreResolver
    uses explicit song/album/artist tags first, SourcePlaylist tags as weaker
    evidence second, and UNKNOWN otherwise. Every genre assignment records
    source and confidence. Pairwise genre similarity normalizes each owner's
    confidence-weighted genre counts into a distribution, then uses weighted
    Jaccard = sum(min(p_i, q_i)) / sum(max(p_i, q_i)).
    """

    metric_key: str = Field(
        description=(
            "Analytics V1 metric key: overview, top_shared_tracks, "
            "pairwise_track_similarity, top_artists, pairwise_artist_similarity, "
            "unique_taste_by_owner, most_similar_pair, most_distinct_pair, "
            "genre_assignments, top_genres, shared_genres, "
            "pairwise_genre_similarity, top_albums, shared_albums, "
            "artist_diversity, or genre_diversity."
        )
    )
    payload: dict[str, Any] = Field(
        description=(
            "Metric payload computed from SourcePlaylist, Contributor, and "
            "UnifiedTrack/MatchedTrack attribution. Overview counts source "
            "playlists, distinct owner_source_id participants, raw tracks, "
            "unique tracks, and tracks with >=2 contributors. Pairwise metrics "
            "use Jaccard = intersection / union. Unique taste ratios use "
            "exclusive_count / owner_total. Pair extremes rank by the mean of "
            "track_jaccard and artist_jaccard and include both components. "
            "V2 payloads include data_coverage and confidence where data may "
            "be incomplete, especially genre metrics. top_genres sums "
            "owner-track confidence weights; pairwise_genre_similarity uses "
            "weighted Jaccard over normalized owner genre distributions; "
            "artist_diversity reports unique_artists, top_artist_share, and "
            "Shannon entropy."
        )
    )
    status: str
    computed_at: str


class ImportPlaybackPayload(BaseModel):
    temp_playlist_id: str
    tracks: list[MatchedTrackItem]


class ImportHistorySourceSummary(BaseModel):
    platform: Literal["netease", "qq"]
    source_playlist_id: str
    title: str
    owner_nickname: str
    import_track_limit: int | None = None
    read_count: int = 0


class ImportHistoryItem(BaseModel):
    session_id: str
    ready_to_play_at: str
    temp_playlist_id: str
    playable_track_count: int
    source_playlists: list[ImportHistorySourceSummary]
    owner_nicknames: list[str]
    created_at: str
    updated_at: str


class ImportSessionDeleteResponse(BaseModel):
    session_id: str
    deleted: bool


class ImportSessionResponse(BaseModel):
    id: str
    status: Literal[
        "pending",
        "reading",
        "ready",
        "partial_failed",
        "failed",
        "previewed",
        "importing",
        "normalizing",
        "matching",
        "syncing_temp",
        "ready_to_play",
    ]
    raw_track_count: int
    source_playlists: list[SourcePlaylistImportResult]
    tracks: list[SourceTrackItem] = []
    created_at: str
    updated_at: str
    failed_stage: Literal[
        "importing",
        "normalizing",
        "matching",
        "syncing_temp",
    ] | None = None
    error: PlaylistPreviewError | None = None
    progress: ImportProgress | None = None
    temp_playlist_id: str | None = None
    ready_to_play_at: str | None = None
    analytics_job_id: str | None = None
    analytics_status: Literal[
        "pending",
        "running",
        "partial",
        "completed",
        "failed",
    ] | None = None
    analytics_results: list[AnalyticsMetric] = []
    matched_tracks: list[MatchedTrackItem] = []
    playback: ImportPlaybackPayload | None = None
